prymer/api/picking.py

```diff
@@ -186,285 +69,112 @@ def score(
 
     # Put it all together
-    return (
-        left_primer.penalty
-        + right_primer.penalty
-        + size_penalty
-        + tm_penalty
-        + left_dist_penalty
-        + right_dist_penalty
-        + left_seq_penalty
-        + right_seq_penalty
-    )
-
-
-def check_primer_overlap(
-    prev_left: Span,
-    prev_right: Span,
-    next_left: Span,
-    next_right: Span,
-    min_difference: int,
-) -> bool:
-    """Check that both (next) primers have at least `min_difference` bases that do not overlap the
-     other (previous) primers.
-
-    Args:
-        prev_left: left primer mapping from the previous primer pair.
-        prev_right: right primer mapping from the previous primer pair
-        next_left: left primer mapping from the current primer pair being compared
-        next_right: right primer mapping from the current primer pair being compared
-        min_difference: the minimum number of bases that must differ between two primers
-
-    Returns:
-        True if the primers are sufficiently different
-    """
-    left = prev_left.length_of_overlap_with(next_left) + min_difference
-    right = prev_right.length_of_overlap_with(next_right) + min_difference
-    return (
-        prev_left.length >= left
-        and next_left.length >= left
-        and prev_right.length >= right
-        and next_right.length >= right
-    )
-
-
-def is_acceptable_primer_pair(primer_pair: PrimerPair, params: FilteringParams) -> bool:
-    """Determine if a primer pair can be kept considering:
-
-    1. the primer pair is in the desired size range
-    2. The primer pair has a melting temperature in the desired range
-
-    Args:
-        primer_pair: the primer pair.
-        params: the parameters used for filtering
-
-    Returns:
-       True if the primer pair passes initial checks and can be considered later on.
-    """
-    return (
-        params.amplicon_sizes.min <= primer_pair.amplicon.length <= params.amplicon_sizes.max
-        and params.amplicon_tms.min <= primer_pair.amplicon_tm <= params.amplicon_tms.max
-    )
+    return left_primer.penalty + right_primer.penalty + size_penalty + tm_penalty
 
 
 def build_primer_pairs(
-    left_primers: Iterable[Oligo],
-    right_primers: Iterable[Oligo],
+    left_primers: list[Oligo],
+    right_primers: list[Oligo],
     target: Span,
-    params: FilteringParams,
-    fasta: FastaFile,
-) -> list[PrimerPair]:
+    amplicon_sizes: MinOptMax[int],
+    amplicon_tms: MinOptMax[float],
+    max_heterodimer_tm: Optional[float],
+    weights: PrimerAndAmpliconWeights,
+    fasta_path: Path,
+) -> Iterator[PrimerPair]:
     """Builds primer pairs from individual left and primers.
 
+    Only primer pairs that meet the requirements for amplicon sizes and tms will be returned.
+    In addition, if `max_heterodimer_tm` is provided then primer pairs with a heterodimer tm
+    exceeding the maximum will also be discarded.
```

Comment:
**question** Why is this necessary when `primer3` will apply these constraints to designed primer pairs if they are specified? e.g. specifying `PRIMER_PRODUCT_MIN_TM` and `PRIMER_PRODUCT_MAX_TM` will constrain the pairs returned by `primer3` to the provided range

Reply:
Because this is used when designing left and right primers independently - so primer3 is not emitting pairs and not cannot check these constraints.